### DISCUSSION TITLE
How to query Enum using Prisma if the FE sends string type?

### DISCUSSION THREAD

**Discussion by kacak9229** [prisma/prisma#12439] [Q&A] [answered]
You can answer this question on SO also if you want

https://stackoverflow.com/questions/71480999/how-to-query-enum-using-prisma

The goal is simple, I want to query enum field in the where clause. Example

prisma version is: "prisma": "3.11.0"

schema.prisma

```
model Todo {
  id Int @default(autoincrement()) @id
  title String
  content String
  status TodoStatus @default(IN_PROGRESS)
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}

enum TodoStatus {
  IN_PROGRESS 
  DONE
}
```
api call

```
app.get('/todos', (req, res){
   const { status } = req.query
   const todos = await prisma.todo.findMany({
      where: { status: status },
      orderBy: {
        id: "asc",
      },
    });

    res.json(todos);
})

```Frontend I use Next.js and it is a select option

example

       ```
     <select
              id="todo"
              name="todo"
              onChange={(e) => setSelectedStatus(e.target.value as Status)}
            >
              {[Status.IN_PROGRESS, Status.DONE].map((status: Status) => {
                return (
                  <option value={status}>{todoStatusToString[status]}</option>
                );
              })}
            </select>
```
The enum value from Next.js

```
export enum Status {
  IN_PROGRESS = "IN_PROGRESS",
  DONE = "DONE",
  ALL = "ALL",
}

export const todoStatusToString = {
  [Status.IN_PROGRESS]: "In progress",
  [Status.DONE]: "Done",
  [Status.ALL]: "All",
};
```
the req.query will be sent from clientside in this format

```
localhost:3000/todos?status=DONE

{ status: "DONE" }
```
or 
```
localhost:3000/todos?status=IN_PROGRESS

{ status: "IN_PROGRESS" } 
```


I know that Prisma is built-in to be typesafe. So my assumption is because the data that we get from the frontend is a string type while the enum on Prisma is looking for either IN_PROGRESS or DONE specifically if we send "DECLINED" to status where clause it will spit the same error.

Any help would be appreciated!

<img width="770" alt="Screenshot 2022-03-19 at 2 02 30 PM" src="https://user-images.githubusercontent.com/2588948/159161893-87a8a782-37bc-4f5d-9f7e-cb6bdfb51227.png">

**Comment by austincrim** [comment, marked as answer]
Hey @kacak9229 👋 ,

You will need to do some kind of type check/guard before using the `status` variable in your Prisma query. 

This will narrow the type of `status` to be `"IN_PROGRESS | "DONE"` which should get rid of the error:
```ts
app.get('/todos', (req, res) => {
 const { status } = req.query
 if (status !== 'IN_PROGRESS" && status !== "DONE") {
   // send error response, log, return from function, etc.
 }
 const todos = await prisma.todo.findMany({
    where: { status: status },
    orderBy: {
      id: "asc",
    },
  });
})
```

Let me know if I can clarify that further!

**Reply by kacak9229** [reply]
Hey Austin, really appreciate your help man! been waiting for days 😄

Can you help me understand why doing `&&` vs `||` will work just fine?

it's like saying if `status` is not equals to `DONE` and `IN_PROGRESS`

I'm a bit confused, and yeah if you have time please help clarify further 😄. Again thank you so much!

**Reply by austincrim** [reply]
You could also write the inverse condition to do the same thing.

```ts
const { status } = req.query;
if (status === 'IN_PROGRESS" || status === "DONE") {
 const todos = await prisma.todo.findMany({
    where: { status: status },
    orderBy: {
      id: "asc",
    },
  });

  res.json(todos);
} else {
  // error stuff here
}
```

**Reply by Zac-Nourish** [reply]
Is there a way to do something like `status in TodoStatus`? That way the type guard doesn't need to be rewritten if the enum ever changes?

**Reply by hevi1991** [reply]
I tried, It works! hope working for you
```
(Object.values(TodoStatus) as string[]).includes(status)
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
